ARSN-298: support `NewerNoncurrentVersions`

alexanderchan-scality opened the pull request:
add support for `NewerNoncurrentVersions` params for
`NoncurrentVersionsExpiration` actions

bert-e commented:
# Hello alexanderchan-scality,



My role is to assist you with the merge of this
pull request. Please type `@bert-e help` to get information
on this process, or consult the [user documentation](
https://bert-e.scality.net/scality/arsenal/doc/user).





*Status report is not available.*

bert-e commented:
# Incorrect fix version



The `Fix Version/s` in issue ARSN-298 contains:


* `7.10.42`



Considering where you are trying to merge, I ignored possible hotfix versions and I expected to find:



* `7.10.43`

* `8.1.84`


Please check the `Fix Version/s` of ARSN-298, or the target
branch of this pull request.

williamlardier commented on lib/models/LifecycleConfiguration.ts:
```diff
@@ -703,14 +706,38 @@ export default class LifecycleConfiguration {
+            actionParams.newerNoncurrentVersions = newerVersionsInt;
+
+        } else {
+            actionParams.newerNoncurrentVersions = 0;
```
If we set `newerNoncurrentVersions` everytime, it may be better not to have it as an optional field, what about

```ts
        const actionParams: {
            error?: ArsenalError;
            days: number;
            newerNoncurrentVersions: number;
        } = { days: 0 };
```
?

williamlardier commented on lib/models/LifecycleConfiguration.ts:
```diff
@@ -852,13 +885,18 @@ export default class LifecycleConfiguration {
-                    entry.addNCVExpiration(days!);
+                    entry.addNCVExpiration('NoncurrentDays', days!);
+
+                    if (newerNoncurrentVersions) {
+                        entry.addNCVExpiration('NewerNoncurrentVersions', newerNoncurrentVersions);
+                    }
```
If the above comment makes sense:

```suggestion
                    entry.addNCVExpiration('NewerNoncurrentVersions', newerNoncurrentVersions!);
```

williamlardier commented on lib/models/LifecycleRule.ts:
```diff
@@ -136,10 +140,18 @@ export default class LifecycleRule {
-    addNCVExpiration(days: number) {
-        this.ncvExpiration = { NoncurrentDays: days };
+    addNCVExpiration(prop: 'NoncurrentDays' | 'NewerNoncurrentVersions', value: number): this;
+    addNCVExpiration(prop: string, value: number ) {
```
Nit:
```suggestion
    addNCVExpiration(prop: string, value: number) {
```

williamlardier reviewed (approved):
LGTM, just some quick comments (I also spent time reading about how this is used by AWS)

bert-e commented:
# Conflict




A conflict has been raised during the creation of
integration branch `w/8.1/feature/ARSN-298/supportNewerNoncurrentVersions` with contents from `feature/ARSN-298/supportNewerNoncurrentVersions`
and `development/8.1`.

**I have not created the integration branch.**


Here are the steps to resolve this conflict:

```bash
 $ git fetch
 $ git checkout -B w/8.1/feature/ARSN-298/supportNewerNoncurrentVersions origin/development/8.1
 $ git merge origin/feature/ARSN-298/supportNewerNoncurrentVersions
 $ # <intense conflict resolution>
 $ git commit
 $ git push -u origin w/8.1/feature/ARSN-298/supportNewerNoncurrentVersions
```

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* 2 peers

jonathan-gramain commented on lib/models/LifecycleRule.ts:
```diff
@@ -136,10 +140,16 @@ export default class LifecycleRule {
+    addNCVExpiration(prop: string, value: number) {
+        const validProps = ['NoncurrentDays', 'NewerNoncurrentVersions'];
+        if (validProps.includes(prop)) {
+            this.ncvExpiration = this.ncvExpiration || { NoncurrentDays: 0 };
```
Are we setting 0 days as a default because the two NCV rules combine with an *and* operator, so an object must match both of them in order to be expired?

alexanderchan-scality replied:
the initial `0` is due to typescript enforcing the field; will  fix this will  this to be an explicit null and a value

jonathan-gramain commented on package.json:
```diff
@@ -3,7 +3,7 @@
     "node": ">=16"
   },
-  "version": "7.10.42",
+  "version": "7.10.43",
```
Could you bump the version in a separate commit? Makes it easier to cherry-pick things later

jonathan-gramain commented on lib/models/LifecycleConfiguration.ts:
```diff
@@ -703,14 +706,41 @@ export default class LifecycleConfiguration {
+            actionParams.days = daysInt;
+        }
+
+        if (subNVExp.NewerNoncurrentVersions) {
```
shouldn't we remove the check at line 703 ` if (!subNVExp.NoncurrentDays)`, now that we introduced a new type of check, or make sure at least one of the two checks is present instead? According to https://docs.aws.amazon.com/AmazonS3/latest/API/API_NoncurrentVersionExpiration.html, it is not required to have `NoncurrentDays` attribute.

alexanderchan-scality replied:
Tested against  AWS S3 and the `NoncurrentDays` field looks to be required. Schema validation would fail otherwise

jonathan-gramain commented on lib/models/LifecycleRule.ts:
```diff
@@ -49,7 +53,7 @@ export default class LifecycleRule {
         }
         if (this.ncvExpiration) {
-            rule.NoncurrentVersionExpiration = this.ncvExpiration;
+            rule.NoncurrentVersionExpiration = this.ncvExpiration
```
probably an unwanted change

jonathan-gramain commented on lib/models/LifecycleConfiguration.ts:
```diff
@@ -703,14 +706,41 @@ export default class LifecycleConfiguration {
+        }
+
+        if (subNVExp.NewerNoncurrentVersions) {
+            const newerVersionsInt = parseInt(subNVExp.NewerNoncurrentVersions[0], 10);
```
Is there a validation earlier that we have a parsable integer value at this point, to avoid getting `NaN` back?

bert-e commented:
# Incorrect fix version



The `Fix Version/s` in issue ARSN-298 contains:


* `7.10.43`

* `8.1.84`



Considering where you are trying to merge, I ignored possible hotfix versions and I expected to find:



* `7.10.43`

* `7.70.0`

* `8.1.84`


Please check the `Fix Version/s` of ARSN-298, or the target
branch of this pull request.

alexanderchan-scality committed:
ARSN-298: support lifecycle NewerNoncurrentVersions

updates `LifecyleConfiguration` and `LifecycleRule` to support the
`NewerNoncurrentVersions` parameter for NoncurrentVersionExpirations

alexanderchan-scality committed:
bump version

bert-e commented:
# History mismatch



Merge commit #0c62b6abff0dc07e1a992e7ad5adda6c1faf2f88 on the integration branch
`w/8.1/feature/ARSN-298/supportNewerNoncurrentVersions` is merging a branch which is neither the current
branch `feature/ARSN-298/supportNewerNoncurrentVersions` nor the development branch
`development/8.1`.

It is likely due to a rebase of the branch `feature/ARSN-298/supportNewerNoncurrentVersions` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**

alexanderchan-scality commented:
/reset

bert-e commented:
# Reset complete



I have successfully deleted this pull request's integration branches.

bert-e commented:
# Conflict




A conflict has been raised during the creation of
integration branch `w/8.1/feature/ARSN-298/supportNewerNoncurrentVersions` with contents from `w/7.70/feature/ARSN-298/supportNewerNoncurrentVersions`
and `development/8.1`.

**I have not created the integration branch.**


Here are the steps to resolve this conflict:

```bash
 $ git fetch
 $ git checkout -B w/8.1/feature/ARSN-298/supportNewerNoncurrentVersions origin/development/8.1
 $ git merge origin/w/7.70/feature/ARSN-298/supportNewerNoncurrentVersions
 $ # <intense conflict resolution>
 $ git commit
 $ git push -u origin w/8.1/feature/ARSN-298/supportNewerNoncurrentVersions
```

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* 2 peers

alexanderchan-scality commented:
/approve

bert-e commented:
# Build failed



The build for [commit](https://github.com/scality/Arsenal/commit/ac53f27bd33a27b8aeb48a51274d2d65405f0445) did not succeed in branch w/8.1/feature/ARSN-298/supportNewerNoncurrentVersions.




*The following options are set:* **approve**

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* 2 peers

alexanderchan-scality commented:
/reset

bert-e commented:
# Reset complete



I have successfully deleted this pull request's integration branches.

bert-e commented:
# Conflict




A conflict has been raised during the creation of
integration branch `w/8.1/feature/ARSN-298/supportNewerNoncurrentVersions` with contents from `w/7.70/feature/ARSN-298/supportNewerNoncurrentVersions`
and `development/8.1`.

**I have not created the integration branch.**


Here are the steps to resolve this conflict:

```bash
 $ git fetch
 $ git checkout -B w/8.1/feature/ARSN-298/supportNewerNoncurrentVersions origin/development/8.1
 $ git merge origin/w/7.70/feature/ARSN-298/supportNewerNoncurrentVersions
 $ # <intense conflict resolution>
 $ git commit
 $ git push -u origin w/8.1/feature/ARSN-298/supportNewerNoncurrentVersions
```

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* 2 peers

alexanderchan-scality commented:
/approve

bert-e commented:
# Unknown command



I didn't understand this comment by @alexanderchan-scality:

> /appove

I don't know what `appove` means.

Please **edit** or **delete** the corresponding comment so I can move on.

bert-e commented:
# Queue out of order



The changeset has received all authorizations to enter the merge queue,
however the merge queue is currently in an incoherent state. In order
to protect this pull request, I have not added the changeset to the
queue.

Look into https://devdocs.scality.net/support/faq/#bert-e for guidance about
what to do next to unlock the situation, or ask for help on the #ci-support
Slack channel.





*The following options are set:* **approve**

alexanderchan-scality commented:
ping

bert-e commented:
# In the queue



The changeset has received all authorizations and has been added to the
relevant queue(s). The queue(s) will be merged in the target development
branch(es) as soon as builds have passed.

The changeset will be merged in:


* :heavy_check_mark: `development/7.10`

* :heavy_check_mark: `development/7.70`

* :heavy_check_mark: `development/8.1`



The following branches will **NOT** be impacted:

* `development/6.4`
* `development/7.4`



There is no action required on your side. You will be notified here once
the changeset has been merged. In the unlikely event that the changeset
fails permanently on the queue, a member of the admin team will
contact you to help resolve the matter.

*IMPORTANT*

Please do not attempt to modify this pull request.

* Any commit you add on the source branch will trigger a new cycle after the
  current queue is merged.
* Any commit you add on one of the integration branches will be **lost**.

If you need this pull request to be removed from the queue, please contact a
member of the admin team now.




*The following options are set:* **approve**

bert-e commented:
# 


I have successfully merged the changeset of this pull request
into targetted development branches:


* :heavy_check_mark: `development/7.10`

* :heavy_check_mark: `development/7.70`

* :heavy_check_mark: `development/8.1`



The following branches have **NOT** changed:

* `development/6.4`
* `development/7.4`



Please check the status of the associated issue ARSN-298.

Goodbye alexanderchan-scality.
